README.md
[![Build Status](http://ci.sagrid.ac.za/buildStatus/icon?job=Fermionic-Molecular-Dynamics-deploy)](http://ci.sagrid.ac.za/job/Fermionic-Molecular-Dynamics-deploy)

# FMD codes (from  T. Neff)

This is a repository of codes NOT CREATED BY ME, but by my advisor, Thomas Neff. They are for FMD calculations.

# Building

## Quickstart

For the impatient:

  1. Edit `Makefile.inc` based on one of the tempates provided, to point to your local compiler and dependencies
  2. Run `make`
  3. SCIENCE !


## Dependencies

The application requires :

  * GCC, gfortran
  * ncurses
  * LAPACK (libblas, liblapack)
  * OpenMPI

# Testing and Delivery

This repository is automatically integrated with [Project CODE-RADE](http://www.africa-grid.org/CODE-RADE). Build, test and deploy scripts are in the [CODE-RADE](CODE-RADE) folder.

  * `build.sh` - this script runs make and builds the static libraries as well as the executable files.
  * `check-build.sh` - this script checks whether the build is successful by running tests
  * `deploy.sh` - this script cleans the build, reconfigures it for deployment (to `/cvmfs`) and then rebuilds with the new target.


doc/CHANGES
commit 5526fd82f99493954bca98320a959b3557909f46
Author: Thomas Neff <t.neff@gsi.de>
Date:   Fri Nov 30 14:23:47 2012 +0100

    Fix creation of git_version.c in Makefile.

commit c45e8ca5f2ab73163682bc79cbde89c469c13efd
Author: Thomas Neff <t.neff@gsi.de>
Date:   Fri Nov 30 13:33:26 2012 +0100

    Fix bug in creation of git_version.c. Add Makefile.inc for Ubuntu quantal.

commit eb4cf7f94d7212c3b6bf50478becb17069b56550
Merge: 0e2a71b 1481bbf
Author: Thomas Neff <t.neff@gsi.de>
Date:   Fri Nov 30 12:35:58 2012 +0100

    Merge branch 'master' of ssh://tplx80.gsi.de/~/var/git-repo-code/fmd
    
    Conflicts:
    	Makefile

commit 0e2a71b65e38a5779a50bc34c1f55031efca52f6
Author: Thomas Neff <t.neff@gsi.de>
Date:   Fri Nov 30 12:05:55 2012 +0100

    Added .gitignore target to Makefile.

commit 1481bbf713fd8c181aa66ed7af18dee10f8e3955
Author: Thomas Neff <t.neff@gsi.de>
Date:   Fri Nov 30 12:05:55 2012 +0100

    Added .gitignore target to Makefile.

commit 328f5ce473761de31e16f064fffae388e2a41110
Author: Thomas Neff <t.neff@gsi.de>
Date:   Fri Nov 30 11:07:01 2012 +0100

    Migration to GIT rcs.

commit 8b48437efd355ca83e468ea5cfc79f5064131c72
Author: Thomas Neff <t.neff@gsi.de>
Date:   Fri Oct 12 04:08:18 2012 +0000

    calcsdradii: calculate radii without center-of-mass corrections

commit bcdf9f1f17eec61be2003eceea01ef49185ae017
Author: Thomas Neff <t.neff@gsi.de>
Date:   Mon Jul 23 11:56:39 2012 +0000

    Make output easier to parse.

commit 2f3ffcca27536e7851ae760beeb81271537c2754
Author: Thomas Neff <t.neff@gsi.de>
Date:   Sun Jun 10 13:56:41 2012 +0000

    Include Omega in shelloccupations output.

commit 8ce4880bd8d0fa79b882c273a5eff9931ec4c397
Author: Thomas Neff <t.neff@gsi.de>
Date:   Tue May 22 08:45:26 2012 +0000

    Remove debugging output.

commit e4a22f49332ed72b2a92b429e71a02a2a1a4e9cd
Author: Thomas Neff <t.neff@gsi.de>
Date:   Tue May 22 08:43:55 2012 +0000

    Bug in MAX macro ...

commit 99bd677f04ddd0c95e028dd29f49864d8051bea5
Author: Thomas Neff <t.neff@gsi.de>
Date:   Tue May 22 07:40:01 2012 +0000

    Two-body densities in momentum space decomposed into partial waves.

commit 732e23dd24d7721e12fb44aade61923e0eb28c6d
Author: Thomas Neff <t.neff@gsi.de>
Date:   Sat May 19 07:34:08 2012 +0000

    Added calctwobodydensitiescoordlme.

commit 2faf6f47a1f489f52d3cd5a3adba87a47194a96e
Author: Thomas Neff <t.neff@gsi.de>
Date:   Sat May 19 07:27:00 2012 +0000

    calctwobodydensitiescoordl calculates two-body densities in different channels.

commit 10f140b52ca43f85b9a1f74870c6958e56cc6dfe
Author: Thomas Neff <t.neff@gsi.de>
Date:   Thu May 3 11:33:20 2012 +0000

    Changed output format for two-nucleon ovlaps.

commit 91339acea068c5cdc708d105d7039746f1784325
Author: Thomas Neff <t.neff@gsi.de>
Date:   Mon Apr 30 15:53:27 2012 +0000

    Bug fix in Y-coordinates, wrong mass was used.

commit af4897f8a95654976d3a684fab6446b7836f25d0
Author: Thomas Neff <t.neff@gsi.de>
Date:   Mon Apr 30 15:27:01 2012 +0000

    Bug fixes for C coordinates, include jacobian in spectroscopic factor calculation.

commit d3d5289cb59493d220b8f83ae7caac4f2f396fd7
Author: Thomas Neff <t.neff@gsi.de>
Date:   Sun Apr 29 14:38:22 2012 +0000

    Added calctwobodydensitiescoordme.

commit 6e96ee008b70cf5d82460aface9616edbbd23328
Author: Thomas Neff <t.neff@gsi.de>
Date:   Sun Apr 29 09:18:04 2012 +0000

    Two-nucleon overlaps in C-coordinates.

commit 20d7207a7fd251c5e5211e4e14cc0c9c04011fb3
Author: Thomas Neff <t.neff@gsi.de>
Date:   Mon Apr 23 19:54:45 2012 +0000

    Fix _estimateangkappa for case of changing nucleon number.

commit b47a6128d453b2afc3df980fdd5f904be327fbc3
Author: Thomas Neff <t.neff@gsi.de>
Date:   Sun Apr 22 08:32:51 2012 +0000

    Fix a bug in Y-coordinates.

commit 0a6e1d466a05969660d4b0b975174114088149e6
Author: Thomas Neff <t.neff@gsi.de>
Date:   Sat Apr 21 11:05:37 2012 +0000

    Bug fixes for Y-coordinates.

commit 2530e3492845cb1d63bec150722941ed3aa19f62
Author: Thomas Neff <t.neff@gsi.de>
Date:   Fri Apr 20 20:02:25 2012 +0000

    OpenMPI with Infiband does not allow fork/popen. Use zlib routines instead
    of popen/gzip.

commit 0b56f9b045e51735bf0a9b30822cf239be83ea81
Author: Thomas Neff <t.neff@gsi.de>
Date:   Fri Apr 20 14:25:52 2012 +0000

    One-nucleon ovlaps.

commit c79b58d7d7a0e6a61f34e433e864f76e08117944
Author: Thomas Neff <t.neff@gsi.de>
Date:   Fri Apr 20 14:21:57 2012 +0000

    One-nucleon ovlaps.

commit f1962cc1a04ac65140ece4b63078f7b7ad727ac1
Author: Thomas Neff <t.neff@gsi.de>
Date:   Fri Apr 20 14:17:20 2012 +0000

    Two-nucleon overlaps in T- and Y-coordinates.

commit 55a71a279006615db721fe9b522d9ffdf85997f1
Author: Thomas Neff <t.neff@gsi.de>
Date:   Sun Apr 15 09:02:12 2012 +0000

    Calculate Two-Nucleon Ovlaps.
    MPI version for calcprojectedMBMEs.

commit d969647721dac8de48cb2222819630ae50ad4ffa
Author: Thomas Neff <t.neff@gsi.de>
Date:   Sun Jan 29 09:38:15 2012 +0000

    Cosmetic: Do no print mpirank, hostname.

commit fe23bb271606c050be69a444285161105afb190e
Author: Thomas Neff <t.neff@gsi.de>
Date:   Tue Nov 29 10:10:03 2011 +0000

    Call dx with '--processors 1' option to work around SMP bug in DataExplorer.

commit 7c0dd96795e6ee81f61cd18e5e3803fa7be5d78a
Author: Thomas Neff <t.neff@gsi.de>
Date:   Tue Nov 15 16:45:21 2011 +0000

    Allow for scaling of matrix elements.

commit 1c8498a10eb12cebe0610c7cd8bf6f5272ddfc35
Author: Thomas Neff <t.neff@gsi.de>
Date:   Tue Oct 4 05:44:48 2011 +0000

    Increase number of eigenstates shown to 15.

commit e3a069796a95b1b9625b26c796664a88c9aea7bc
Author: Thomas Neff <t.neff@gsi.de>
Date:   Thu Sep 15 08:10:24 2011 +0000

    Calculate dimensionless deformation parameter beta.

commit 3804e1843c2c56082ab24fc3f02a1926d7868217
Author: Thomas Neff <t.neff@gsi.de>
Date:   Fri Jul 29 09:53:58 2011 +0000

    Bug fix for HF parity calculation.

commit 3540a1372ee4027d5c4917f5b53e690609ef13f4
Author: Thomas Neff <t.neff@gsi.de>
Date:   Fri Jul 29 08:55:08 2011 +0000


fmd/NOsci.c
/**

  \file NOsci.c 

  Harmonic Oscillator quanta

  (c) 2007 Thomas Neff

*/

#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <complex.h>

#include "Gaussian.h"
#include "SlaterDet.h"
#include "Projection.h"
#include "NOsci.h"

#include "numerics/cmath.h"
#include "misc/utils.h"
#include "misc/physics.h"


// NOsci
ManyBodyOperator OpNOsci = {
  name : "NOsci",
  rank : 0,
  pi : 0,
  dim : 2,
  size : 2,
  par : NULL,
  me : calcNOsciod
};


// 0hbw oscillator quanta

int nq[] = {  0,                                 // no nucleons
              0, 0,                              // s-shell
              1, 2, 3, 4, 5, 6,                  // p-shell
              8,10,12,14,16,18,20,22,24,26,      // sd-shell
              29,32,35,38,41,44,47,50,53,56,     // pf-shell
              59,62,65,68,71,74,77,80,83,86 };

static inline double nquanta0(const SlaterDet* Q)
{
  return (1.5*Q->A + nq[Q->Z] + nq[Q->N]);
}

// calculate x^2 and p^2 matrix elements
static void ob_nosci(void* par,
                     const Gaussian*G1, const Gaussian* G2, 
                     const GaussianAux* X, 
                     complex double nosci[2])
{
  // assume Slater determinants are in origin 
  double Xcm[3] = {0, 0, 0};
  double Vcm[3] = {0, 0, 0};

  int i;
  complex double rho2cm, pi2cm;

  rho2cm = 3*X->alpha;
  for (i=0; i<3; i++)
    rho2cm += csqr(X->rho[i]-Xcm[i]);

  pi2cm = 3*X->lambda;
  for (i=0; i<3; i++)
    pi2cm += csqr(X->pi[i]-mass(G1->xi)*Vcm[i]);
  
  nosci[0] = rho2cm* X->Q;
  nosci[1] = pi2cm* X->Q;
}


void calcNOsciod(void* Par,
                 const SlaterDet* Q, const SlaterDet* Qp,
                 const SlaterDetAux* X, 
                 complex double nosci[2])
{
  OneBodyOperator op_ob_nosci = {dim: 2, opt: 1, par: NULL, me: ob_nosci};

  calcSlaterDetOBMEod(Q, Qp, X, &op_ob_nosci, nosci);
}


void writeprojectedNOsci(FILE* fp,
                         const Projection* P,
                         const complex double (**nosci)[2],
                         const Eigenstates* E)
{
  int odd=P->odd;
  int jmax=P->jmax;

  int p,j,i;

  char prefix[8];

  int* idx;
  int ngood;
  complex double* norm;
  complex double *H, (*NO)[2];
  double x2, p2;

  for (p=0; p<=1; p++)
    for (j=odd; j<jmax; j=j+2) {
      
      ngood = E->ngood[idxpij(jmax,p,j)];

      if (ngood) {

	if(odd) sprintf(prefix, "[%d/2%c]", j, p ? '-' : '+'); 
	else    sprintf(prefix, "[%d%c]", j/2, p ? '-' : '+'); 

	idx = E->index[idxpij(jmax,p,j)];
	norm = E->norm[idxpij(jmax,p,j)];
	H = E->v[idxpij(jmax,p,j)];
 
	NO = nosci[idxpij(jmax,p,j)];

	fprintf(fp, "\n%s  N     = ", prefix); 
	for (i=0; i<ngood; i++) 
	  fprintf(fp, "   %8.5f", creal(norm[idx[i]]));
	fprintf(fp, "\n%s  H     = ", prefix); 
	for (i=0; i<ngood; i++) 
	  fprintf(fp, "   %8.3f", hbc*creal(H[idx[i]]));
	fprintf(fp, "\n%s  NOsci = ", prefix); 
	for (i=0; i<ngood; i++) {
          x2 = creal(NO[idx[i]][0]/norm[idx[i]]);
          p2 = creal(NO[idx[i]][1]/norm[idx[i]]);
	  fprintf(fp, "   %8.3f", sqrt(x2*p2));
        }
	fprintf(fp, "\n");	  
      }
    }
}
